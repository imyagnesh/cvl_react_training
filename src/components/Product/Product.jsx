import React, { memo } from 'react';
import Review from '../Review';

function Product({
  id,
  image,
  title,
  description,
  price,
  rating,
  addToCart,
  cartItem,
  updateCartItem,
  deleteCartItem,
  isAdding,
  isUpdating,
  isDeleting,
}) {
  return (
    <div className="max-w-7xl mx-auto px-2 sm:px-6 lg:px-8 py-4">
      <div className="w-full grid grid-cols-1 gap-y-8 gap-x-6 items-start sm:grid-cols-12 lg:gap-x-8">
        <div className="aspect-w-2 aspect-h-3 rounded-lg bg-gray-100 overflow-hidden sm:col-span-4 lg:col-span-2">
          <img src={image} alt={title} className="object-center object-cover" />
        </div>
        <div className="sm:col-span-8 lg:col-span-10">
          <h2 className="text-2xl font-extrabold text-gray-900 sm:pr-12">
            {title}
          </h2>

          <section aria-labelledby="information-heading" className="mt-2">
            <h3 id="information-heading">{description}</h3>

            <p className="text-2xl text-gray-900">
              {new Intl.NumberFormat('en-IN', {
                style: 'currency',
                currency: 'INR',
              }).format(price)}
            </p>

            {/* Reviews */}
            <Review {...rating} />
          </section>

          {cartItem ? (
            <div className="flex mt-6 items-center">
              <button
                type="button"
                disabled={isUpdating || isDeleting}
                onClick={() =>
                  updateCartItem({
                    ...cartItem,
                    quantity: cartItem.quantity + 1,
                  })
                }
                className="w-full bg-indigo-600 border border-transparent rounded-md py-3 px-8 flex items-center justify-center text-base font-medium text-white hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500 disabled:bg-slate-400"
              >
                +
              </button>
              <span className="mx-4 text-2xl font-semibold">
                {cartItem.quantity}
              </span>
              <button
                type="button"
                disabled={isUpdating || isDeleting}
                onClick={() => {
                  if (cartItem.quantity > 1) {
                    updateCartItem({
                      ...cartItem,
                      quantity: cartItem.quantity - 1,
                    });
                  } else {
                    deleteCartItem(cartItem);
                  }
                }}
                className="w-full bg-indigo-600 border border-transparent rounded-md py-3 px-8 flex items-center justify-center text-base font-medium text-white hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500 disabled:bg-slate-400"
              >
                -
              </button>
            </div>
          ) : (
            <button
              type="button"
              disabled={isAdding}
              onClick={() => addToCart(id)}
              className="mt-6 w-full bg-indigo-600 border border-transparent rounded-md py-3 px-8 flex items-center justify-center text-base font-medium text-white hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500 disabled:bg-slate-400"
            >
              Add to bag
            </button>
          )}
        </div>
      </div>
    </div>
  );
}

export default memo(Product);
