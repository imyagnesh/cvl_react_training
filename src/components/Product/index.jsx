import React from 'react';
import axiosInstance from '../../utils/axiosInstance';
import Review from '../Review';

function Product({ data, addToCart, cartItem }) {
  return (
    <div className="max-w-7xl mx-auto px-2 sm:px-6 lg:px-8 py-4">
      <div className="w-full grid grid-cols-1 gap-y-8 gap-x-6 items-start sm:grid-cols-12 lg:gap-x-8">
        <div className="aspect-w-2 aspect-h-3 rounded-lg bg-gray-100 overflow-hidden sm:col-span-4 lg:col-span-2">
          <img
            src={data.image}
            alt={data.title}
            className="object-center object-cover"
          />
        </div>
        <div className="sm:col-span-8 lg:col-span-10">
          <h2 className="text-2xl font-extrabold text-gray-900 sm:pr-12">
            {data.title}
          </h2>

          <section aria-labelledby="information-heading" className="mt-2">
            <h3 id="information-heading">{data.description}</h3>

            <p className="text-2xl text-gray-900">
              {new Intl.NumberFormat('en-IN', {
                style: 'currency',
                currency: 'INR',
              }).format(data.price)}
            </p>

            {/* Reviews */}
            <Review {...data.rating} />
          </section>

          {cartItem ? (
            <div className="flex mt-6 items-center">
              <button
                type="button"
                onClick={() => addToCart(data.id)}
                className="w-full bg-indigo-600 border border-transparent rounded-md py-3 px-8 flex items-center justify-center text-base font-medium text-white hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500"
              >
                +
              </button>
              <span className="mx-4 text-2xl font-semibold">
                {cartItem.quantity}
              </span>
              <button
                type="button"
                onClick={() => addToCart(data.id)}
                className="w-full bg-indigo-600 border border-transparent rounded-md py-3 px-8 flex items-center justify-center text-base font-medium text-white hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500"
              >
                -
              </button>
            </div>
          ) : (
            <button
              type="button"
              onClick={() => addToCart(data.id)}
              className="mt-6 w-full bg-indigo-600 border border-transparent rounded-md py-3 px-8 flex items-center justify-center text-base font-medium text-white hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500"
            >
              Add to bag
            </button>
          )}
        </div>
      </div>
    </div>
  );
}

export default Product;
