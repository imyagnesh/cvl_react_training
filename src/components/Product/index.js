import { connect } from 'react-redux';
import Product from './Product';

const mapStateToProps = (state, props) => ({
  cartItem: state.cart.find((x) => x.productId === props.id),
  isAdding: state.loading.find(
    (x) => x.loadingId === props.id && x.actionName === 'ADD_CART'
  ),
  isUpdating: state.loading.find(
    (x) => x.loadingId === props.id && x.actionName === 'UPDATE_CART'
  ),
  isDeleting: state.loading.find(
    (x) => x.loadingId === props.id && x.actionName === 'DELETE_CART'
  ),
});

const mapDispatchToProps = (dispatch) => ({
  addToCart: (productId) =>
    dispatch({
      type: 'ADD_CART_REQUEST',
      payload: { loadingId: productId, message: 'Adding Item to cart' },
    }),
  updateCartItem: (cartItem) =>
    dispatch({
      type: 'UPDATE_CART_REQUEST',
      payload: {
        cartItem,
        message: 'Updating Item to cart',
      },
    }),
  deleteCartItem: (cartItem) =>
    dispatch({
      type: 'DELETE_CART_REQUEST',
      payload: {
        cartItem,
        message: 'Updating Item to cart',
      },
    }),
});

export default connect(mapStateToProps, mapDispatchToProps)(Product);
