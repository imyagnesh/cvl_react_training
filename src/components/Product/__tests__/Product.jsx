import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import Product from '../Product';

describe('Should render product Component', () => {
  const addToCart = jest.fn();
  const updateCartItem = jest.fn();
  const deleteCartItem = jest.fn();

  const props = {
    id: 1,
    title: 'Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops',
    price: 109.95,
    description:
      'Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday',
    category: "men's clothing",
    image: 'https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg',
    rating: {
      rate: 3.9,
      count: 120,
    },
    cartItem: undefined,
    addToCart,
    updateCartItem,
    deleteCartItem,
    isAdding: false,
    isUpdating: false,
    isDeleting: false,
  };

  beforeEach(() => {
    render(<Product {...props} />);
  });

  test('should take screenshot', () => {
    const { container } = render(<Product {...props} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('should display add to cart button', () => {
    const btn = screen.getByRole('button');
    expect(btn.innerHTML).toBe('Add to bag');
    expect(btn).not.toBeDisabled();
    fireEvent.click(btn);
    expect(addToCart).toBeCalled();
    expect(addToCart).toBeCalledWith(1);
  });
});
