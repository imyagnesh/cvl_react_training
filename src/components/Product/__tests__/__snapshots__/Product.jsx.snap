// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render product Component should take screenshot 1`] = `
<div
  class="max-w-7xl mx-auto px-2 sm:px-6 lg:px-8 py-4"
>
  <div
    class="w-full grid grid-cols-1 gap-y-8 gap-x-6 items-start sm:grid-cols-12 lg:gap-x-8"
  >
    <div
      class="aspect-w-2 aspect-h-3 rounded-lg bg-gray-100 overflow-hidden sm:col-span-4 lg:col-span-2"
    >
      <img
        alt="Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops"
        class="object-center object-cover"
        src="https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg"
      />
    </div>
    <div
      class="sm:col-span-8 lg:col-span-10"
    >
      <h2
        class="text-2xl font-extrabold text-gray-900 sm:pr-12"
      >
        Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops
      </h2>
      <section
        aria-labelledby="information-heading"
        class="mt-2"
      >
        <h3
          id="information-heading"
        >
          Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday
        </h3>
        <p
          class="text-2xl text-gray-900"
        >
          â‚¹109.95
        </p>
        <div
          class="mt-6"
        >
          <h4
            class="sr-only"
          >
            Reviews
          </h4>
          <div
            class="flex items-center"
          >
            <div
              class="flex items-center"
            >
              <svg
                aria-hidden="true"
                class="text-gray-900 h-5 w-5 flex-shrink-0"
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M9.049 2.927c.3-.921 1.603-.921 1.902 0l1.07 3.292a1 1 0 00.95.69h3.462c.969 0 1.371 1.24.588 1.81l-2.8 2.034a1 1 0 00-.364 1.118l1.07 3.292c.3.921-.755 1.688-1.54 1.118l-2.8-2.034a1 1 0 00-1.175 0l-2.8 2.034c-.784.57-1.838-.197-1.539-1.118l1.07-3.292a1 1 0 00-.364-1.118L2.98 8.72c-.783-.57-.38-1.81.588-1.81h3.461a1 1 0 00.951-.69l1.07-3.292z"
                />
              </svg>
              <svg
                aria-hidden="true"
                class="text-gray-900 h-5 w-5 flex-shrink-0"
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M9.049 2.927c.3-.921 1.603-.921 1.902 0l1.07 3.292a1 1 0 00.95.69h3.462c.969 0 1.371 1.24.588 1.81l-2.8 2.034a1 1 0 00-.364 1.118l1.07 3.292c.3.921-.755 1.688-1.54 1.118l-2.8-2.034a1 1 0 00-1.175 0l-2.8 2.034c-.784.57-1.838-.197-1.539-1.118l1.07-3.292a1 1 0 00-.364-1.118L2.98 8.72c-.783-.57-.38-1.81.588-1.81h3.461a1 1 0 00.951-.69l1.07-3.292z"
                />
              </svg>
              <svg
                aria-hidden="true"
                class="text-gray-900 h-5 w-5 flex-shrink-0"
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M9.049 2.927c.3-.921 1.603-.921 1.902 0l1.07 3.292a1 1 0 00.95.69h3.462c.969 0 1.371 1.24.588 1.81l-2.8 2.034a1 1 0 00-.364 1.118l1.07 3.292c.3.921-.755 1.688-1.54 1.118l-2.8-2.034a1 1 0 00-1.175 0l-2.8 2.034c-.784.57-1.838-.197-1.539-1.118l1.07-3.292a1 1 0 00-.364-1.118L2.98 8.72c-.783-.57-.38-1.81.588-1.81h3.461a1 1 0 00.951-.69l1.07-3.292z"
                />
              </svg>
              <svg
                aria-hidden="true"
                class="text-gray-900 h-5 w-5 flex-shrink-0"
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M9.049 2.927c.3-.921 1.603-.921 1.902 0l1.07 3.292a1 1 0 00.95.69h3.462c.969 0 1.371 1.24.588 1.81l-2.8 2.034a1 1 0 00-.364 1.118l1.07 3.292c.3.921-.755 1.688-1.54 1.118l-2.8-2.034a1 1 0 00-1.175 0l-2.8 2.034c-.784.57-1.838-.197-1.539-1.118l1.07-3.292a1 1 0 00-.364-1.118L2.98 8.72c-.783-.57-.38-1.81.588-1.81h3.461a1 1 0 00.951-.69l1.07-3.292z"
                />
              </svg>
              <svg
                aria-hidden="true"
                class="text-gray-200 h-5 w-5 flex-shrink-0"
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M9.049 2.927c.3-.921 1.603-.921 1.902 0l1.07 3.292a1 1 0 00.95.69h3.462c.969 0 1.371 1.24.588 1.81l-2.8 2.034a1 1 0 00-.364 1.118l1.07 3.292c.3.921-.755 1.688-1.54 1.118l-2.8-2.034a1 1 0 00-1.175 0l-2.8 2.034c-.784.57-1.838-.197-1.539-1.118l1.07-3.292a1 1 0 00-.364-1.118L2.98 8.72c-.783-.57-.38-1.81.588-1.81h3.461a1 1 0 00.951-.69l1.07-3.292z"
                />
              </svg>
            </div>
            <p
              class="sr-only"
            >
              3.9
               out of 5 stars
            </p>
            120
             reviews
          </div>
        </div>
      </section>
      <button
        class="mt-6 w-full bg-indigo-600 border border-transparent rounded-md py-3 px-8 flex items-center justify-center text-base font-medium text-white hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500 disabled:bg-slate-400"
        type="button"
      >
        Add to bag
      </button>
    </div>
  </div>
</div>
`;
